Sources/PenguinStructures/FixedSizeArray.swift

```diff
@@ -0,0 +1,269 @@
+  // Note: we don't have a generalization to `Sequence` because we couldn't
+  // find an implementation optimizes nearly as well, and in practice
+  // `Sequence`'s that are not `Collection`s are extremely rare.
+
+  /// Creates an instance containing the elements of `source` except the one
+  /// at `targetPosition`.
+  ///
+  /// Requires: `source.indices.contains(targetPosition)`
+  init(_ source: ArrayN<Self>, removingAt targetPosition: Index)
+  
+  /// Returns a fixed-sized collection containing the same elements as `self`,
+  /// with `newElement` inserted at `targetPosition`.
+  func inserting(_ newElement: Element, at targetPosition: Index) -> ArrayN<Self>
+}
+
+/// Default implementation of `CustomStringConvertible` conformance.
+public extension FixedSizeArray {
+  var description: String { "\(Array(self))"}
+  
+  @_transparent
+  func withUnsafeBufferPointer<R>(
+      _ body: (UnsafeBufferPointer<Element>) throws -> R
+  ) rethrows -> R {
+    let count = self.count
+    return try withUnsafePointer(to: self) { p in
+      try body(
+          UnsafeBufferPointer<Element>(
+              start: UnsafeRawPointer(p)
+                  .assumingMemoryBound(to: Element.self),
+              count: count))
+    }
+  }
+
+  @_transparent
+  mutating func withUnsafeMutableBufferPointer<R>(
+      _ body: (UnsafeMutableBufferPointer<Element>) throws -> R
+  ) rethrows -> R {
+    let count = self.count
+    return try withUnsafeMutablePointer(to: &self) { p in
+      try body(
+          UnsafeMutableBufferPointer<Element>(
+              start: UnsafeMutableRawPointer(p)
+                  .assumingMemoryBound(to: Element.self),
+              count: count))
+    }
+  }
+
+  /// Returns a fixed-sized collection containing the same elements as `self`,
+  /// with `newElement` inserted at the start.
+  func prepending(_ newElement: Element) -> ArrayN<Self> {
+    .init(head: newElement, tail: self)
+  }
+}
+
+/// A fixed sized collection of 0 elements.
+public struct Array0<T> : FixedSizeArray {
+  public init() {  }
+  
+  /// Creates an instance containing exactly the elements of `source`.
+  ///
+  /// Requires: `source.isEmpty`
+  @inline(__always)
+  public init<Source: Collection>(_ source: Source)
+      where Source.Element == Element
+  {
+    precondition(source.isEmpty, "Too many elements in source")
+  }
+  
+  /// Creates an instance containing the elements of `source` except the one
+  /// at `targetPosition`.
+  ///
+  /// Requires: `source.indices.contains(targetPosition)`
+  public init(_ source: ArrayN<Self>, removingAt targetPosition: Index) {
+    precondition(targetPosition == 0, "Index out of range.")
+    self.init()
+  }
+
+  /// Returns a fixed-sized collection containing the same elements as `self`,
+  /// with `newElement` inserted at `targetPosition`.
+  public func inserting(_ newElement: Element, at i: Index) -> ArrayN<Self> {
+    precondition(i == 0, "Index out of range.")
```

Comment:
Ditto for error message here & below.

Reply:
https://github.com/saeta/penguin/pull/24#discussion_r428787052